src/gui/touchscreengui.cpp

```diff
@@ -1124,3 +1051,110 @@ void TouchScreenGUI::show()
 
 	setVisible(true);
 }
+
+v2s32 TouchScreenGUI::getPointerPos() {
+	if (m_draw_crosshair)
+		return v2s32(m_screensize.X / 2, m_screensize.Y / 2);
+	return m_pointer_pos[m_move_id];
+}
+
+void TouchScreenGUI::emitMouseEvent(EMOUSE_INPUT_EVENT type) {
+	v2s32 pointer_pos = getPointerPos();
+
+	SEvent event{};
+	event.EventType               = EET_MOUSE_INPUT_EVENT;
+	event.MouseInput.X            = pointer_pos.X;
+	event.MouseInput.Y            = pointer_pos.Y;
+	event.MouseInput.Shift        = false;
+	event.MouseInput.Control      = false;
+	event.MouseInput.ButtonStates = 0;
+	event.MouseInput.Event        = type;
+	m_receiver->OnEvent(event);
+}
+
+void TouchScreenGUI::applyContextControls(const PointedThing &pointed, const TouchControlHint &hint) {
+	// Since the pointed thing has already been determined when this function
+	// is called, we cannot use this function to update the shootline.
+
+	TouchControlMode mode;
+	switch (pointed.type) {
+	case POINTEDTHING_NOTHING:
+		mode = hint.pointed_nothing;
+		break;
+	case POINTEDTHING_NODE:
+		mode = hint.pointed_node;
+		break;
+	case POINTEDTHING_OBJECT:
+		mode = hint.pointed_object;
+		break;
+	}
```

Comment:
for extra good class conception (is this a word?) there could be a method in TouchControlHint that chooses and returns the mode based on the `PointedThingType`